source/OrderedTypes/ZornsLemma.lagda

```diff
@@ -0,0 +1,375 @@
+open import UF.Univalence
+open import UF.FunExt
+open import UF.UA-FunExt
+open import Ordinals.Underlying
+open import Ordinals.Equivalence
+open import UF.Logic
+open import UF.Base
+open import UF.Equiv
+open import UF.Subsingletons
+open import UF.Choice
+open import UF.Subsingletons-FunExt
+open import UF.EquivalenceExamples
+open import UF.UniverseEmbedding
+
+module OrderedTypes.ZornsLemma
+ {𝓤 : Universe}
+ {𝓣 : Universe}
+ (pt : propositional-truncations-exist)
+ (ua : Univalence)
+ {X : 𝓤 ̇}
+ (_≪_ : X → X → 𝓣 ̇)
+  where
+
+open PropositionalTruncation pt
+open import Ordinals.OrdinalOfOrdinals ua
+open import Ordinals.BuraliForti ua
+open import UF.ImageAndSurjection pt
+
+fe : FunExt
+fe = Univalence-gives-FunExt ua
+pe : Prop-Ext
+pe = Univalence-gives-Prop-Ext ua
+pe' : PropExt
+pe' 𝓤 = pe {𝓤}
+fe' : Fun-Ext
+fe' {𝓤} {𝓥} = fe 𝓤 𝓥
+
+open UF.Choice.choice-functions pt pe' fe
+open UF.Choice.Univalent-Choice fe pt
+open import Ordinals.Arithmetic fe
+open import OrderedTypes.Poset fe'
+open import Ordinals.WellOrderingTaboo fe' pe
+open PosetAxioms
+open ClassicalWellOrder pt
+open UF.Logic.AllCombinators pt fe'
+open UF.Choice.ExcludedMiddle pt fe
+open inhabited-subsets pt
+
+_⋘_ : X → X → (𝓤 ⊔ 𝓣) ̇
+a ⋘ b =  a ≪ b × (a ≠ b)
+
+is-chain : {𝓥 : Universe} → (Y : 𝓟 {𝓥} X ) →  (𝓤 ⊔ 𝓣 ⊔ 𝓥) ̇
+is-chain Y = (x y : X) → x ∈ Y → y ∈ Y → ∥ (x ≪ y + y ≪ x) ∥
+
+has-maximal-element-strong :  (𝓤 ⊔ 𝓣) ̇
+has-maximal-element-strong = Σ x ꞉ X , ((y : X) → x ≪ y → x ＝ y)
+
+all-chains-have-upper-bound : {𝓥 : Universe} → (𝓤 ⊔ 𝓣 ⊔ (𝓥 ⁺))  ̇
+all-chains-have-upper-bound {𝓥} =
+ (Y : 𝓟 {𝓥} X) → (is-chain Y) → Σ x ꞉ X , (∀ y → y ∈ Y → y ≪ x)
+
+\end{code}
+
+The following is not a direct formalization of any extant proof,
+but follows the same proof idea as the standard transfinite-induction-based
+proof of Zorn's lemma. The closest informal analog to the proof can be found
+here: https://math.stackexchange.com/a/2889205 .
+
+We do a proof by contradiction (which is why we have that double negation),
+where we assume that even though all chains have a least upper bound,
+the set at hand does not have a maximal element.
+These assumptions allow us to construct a function `f' from the ordinals to X,
+which is a set, which preserves order, is therefore left-cancellable, and again
+in turn an equivalence.
+`f' being order preserving relies on the fact that there is no maximal element,
+as otherwise `f' would (as defined) map many ordinals to the maximal element.
+An equivalence between the ordinals and a set is impossible,
+so there must (not-not) be a maximum.
+
+While this is the core of the proof logic, the final result to be used elsewhere
+is found below, and takes in more reasonable inputs.
+
+\begin{code}
+
+choice-function-gives-not-not-zorns-lemma :
+ Excluded-Middle →
+ poset-axioms (_≪_) →
+ all-chains-have-upper-bound  →
+ (Σ ε ꞉ (𝓟 {𝓣 ⊔ 𝓤} X → X) , ((A : 𝓟 X) → is-inhabited A → ε A ∈ A)) →
+ ¬¬ has-maximal-element-strong
```

Comment:
Given that you are assuming excluded middle, you could use `DNE` and `EM-gives-DNE` to conclude `has-maximal-element-strong` without `¬¬`.

Reply:
I think to get that I would have to weaken `maximal-element-strong` to `maximal-element` (like I do later) since I don't think maximal-element-strong is a proposition. The other problem I had there as needing `¬ has-maximal-element-strong` to get the contradiction to begin with --- right now I'm taking that as an argument to the top-level function which I think looks a bit better, since otherwise all of the terms need to be parametrized by `¬ has-maximal-element strong` but the change I just pushed changes it to make the term that's being constructed of type `¬ has-maximal-element strong →   𝟘`.